### PR TITLE
Added support for 'length' attribute for plugins.

### PR DESCRIPTION
**Pull request by taylorzane**
This feature allows one to truncate the line length of any line (title or menu bar item) to a specified length, per line.

I have also implemented a default line length of 20 characters, to prevent scripts from outputting an unnecessarily long title, or menu bar item, such as in #124.

One may override the default line length by using the `length` attribute.

Example: 

```bash
'This is a very long string of text that needs to be longer than 20 characters | length=100'
||
'This is a very long string that can be any length so we will disable the truncation | length=-1'
```

I will also have to submit a PR for the plugins repo to list this attribute.

### CODE REVIEW

**Comment by johngibb** [comment]
Why not just truncate in your shell script?

    $ echo "this is a long string" | cut -c1-10
    this is a

**Comment by jcaille** [comment]
That's nice !

Should we provide a way to see the full text though ? Maybe by hovering over the menu item ?
What happens if we pass `length=0` ? Could that make is so that the text is not truncated ?

File: App/BitBar/Plugin.m [hunk 1]
```diff
@@ -36,7 +36,13 @@ - (NSMenuItem*) buildMenuItemWithParams:(NSDictionary *)params {
     return nil;
   }
   
-  NSString * title = [params objectForKey:@"title"];
+  NSString * fullTitle = params[@"title"];
+
+  CGFloat lineLength = params[@"length"] ? [params[@"length"] floatValue] : 20;
```
Comment (0wnrepo) [inline App/BitBar/Plugin.m]: you could define these values somewhere, so they are not hardcoded

File: App/BitBar/Plugin.m [hunk 2]
```diff
@@ -55,7 +61,13 @@ - (NSMenuItem*) buildMenuItemWithParams:(NSDictionary *)params {
 
 - (NSAttributedString*) attributedTitleWithParams:(NSDictionary *)params {
 
-  NSString * title = params[@"title"];
+  NSString * fullTitle = params[@"title"];
+
+  CGFloat lineLength = params[@"length"] ? [params[@"length"] floatValue] : 20;
```
Comment (0wnrepo) [inline App/BitBar/Plugin.m]: same here

**Comment by taylorzane** [comment]
@jcaille I figure some people may have a need for a zero-length string, so making `length=0` would prevent users from doing that.

We definitely need a way to view the full length though. I think that hovering is our safest bet. Scrolling would be tough because our script should ideally be stateless,

@0wnrepo You're probably right. Especially if BitBar ever gets a config file, and this ends up as one of the config variables.

**Comment by taylorzane** [comment]
@jcaille So having a tooltip provide the full string doesn't look terrible. I was also thinking that we could do what most native menu's do, and provide more detail (i.e. the full string) when you press and hold the `option` key. 

This is a 100 character tooltip truncated to the default 20 characters:
screen shot 2016-01-07 at 08 58 52

This is a 300 character tooltip truncated to the default 20 characters:
screen shot 2016-01-07 at 09 04 23

**Comment by matryer** [comment]
When it's truncated, we should add `…`, don't you think?

**Comment by taylorzane** [comment]
Heh, probably. Otherwise people may not realize why their output is missing.

**Comment by taylorzane** [comment]
So the option key modifier works well, as long as your string isn't ridiculously long...It seems like anything over ~200 characters gets automatically truncated. So we may be better off with the tooltip, obviously it's up to the community to decide.

No `option` modifier:
screen shot 2016-01-07 at 10 12 50

With `option` modifier:
screen shot 2016-01-07 at 10 14 18

(Also, I implemented the ellipsis per request.)

**Comment by matryer** [comment]
I think the tooltip is a good idea - maybe we can push the default limit up a bit?

We should keep the `option` alternative as something people can provide for themselves later. That'd be cool.

**Comment by taylorzane** [comment]
I agree. I've found, even with some simple scripts, that 20 characters is a bit too small. Are you referring to passing an `option` param after the string? Or were you just referring to people who want that, can implement it themselves?

**Comment by matryer** [comment]
(see my edit)

**Comment by matryer** [comment]
Yes. But that interface might not work - because it's essentially a different option completely. Needs some thought.

**Comment by jcaille** [comment]
I think the `length` keyword is a good idea, but I would personally remove the default for the following reason 

* It does not encourage plugin writer to be succinct.
* The cut might be weirdly placed, or cut off important information.
* It might be seen as a bug by new developers who might wonder why their text is being truncated (and maintaining option is not easily discoverable)
* It might break existing plugins
* Bitbar can display errors inline (allowing to easily see what went wrong). Truncating those errors make it less obvious.

Just keeping the `length` parameter available would allow plugins that might output very long text to toggle the option.

**Comment by matryer** [comment]
@jcaille makes some good points.

**Comment by taylorzane** [comment]
Fair enough. I didn't consider the issue with errors. But being considered a bug by anyone is definitely not ideal. That's good enough of an argument for me. I'll kill the default, and implement the tooltips.

**Comment by taylorzane** [comment]
I pushed a new commit. It includes the following:
- Default limit is gone. You must specify `length=some_positive_integer` for any truncation to occur.
  - `length=0` will simply leave an ellipsis with a tooltip of the full message.
- An ellipsis will be appended to any truncated string.
- A tooltip will be added to any truncated menu item, that will display the full string.

I tested this param with multiple different cases:
- [x] With params `color`, `size`, `font` (all possible combinations)
- [x] With param `bash`
- [x] With param `dropdown=false`

All mentioned tests worked flawlessly.

**Comment by matryer** [comment]
Excellent - I wonder if you could update the README too?

**Comment by taylorzane** [comment]
Absolutely. I didn't realize that the API section was moved here.

**Comment by matryer** [comment]
Nice work :)

**Comment by matryer** [comment]
... while you're in there, could you stick this link on the shebang word? https://en.wikipedia.org/wiki/Shebang_(Unix)

**Comment by taylorzane** [comment]
Haha certainly

**Comment by taylorzane** [comment]
Alright, README has been updated with the shebang link and the necessary `length` documentation.

**Comment by matryer** [comment]
Excellent. Nice work. Thanks to everyone for their input.

**Comment by alexrockt** [comment]
Is there a way to pad a string using the length parameter?

The output from the ping plugin has different widths, depending on the ping time - I would love to have a fixed width - say 10 characters so that the output doesn't "jump" when it's getting longer/shorter. Or is there a way to do it with bash? I tried using printf (printf "%10s" "$MSG") but this didn't work. It works in the terminal but somehow the leading spaces are removed.

**Comment by iosdeveloper** [comment]
@mm2270 Please open a new issue.

**Comment by mm2270** [comment]
Will do. Thanks.

**Pull request merged**